test/integration/apiserver/flowcontrol/concurrency_util_test.go

```diff
@@ -0,0 +1,506 @@
+// of concurrency. In order to mitigate the effects of system noise, authorization webhook is used to artificially
+// increase request execution time to make the system noise relatively insignificant.
+// This test calculates the server-side observed concurrency from average priority level seat utilization APF metric.
+// It also assumes that
+// (server-side request execution throughput) == (client-side request throughput) and derives a formula to
+// calculate the client-side expected concurrency. The two are compared and a small error bound is determined
+// from estimating the noise using 2*(standard deviation of requenst latency)/(avg request latency).
+func TestConcurrencyIsolation(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.APIPriorityAndFairness, true)()
+	// NOTE: disabling the feature should fail the test
+	// start webhook server
+	serv := &mockV1Service{allow: true, statusCode: 200}
+	s, err := NewV1TestServer(serv, testcerts.ServerCert, testcerts.ServerKey, testcerts.CACert)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer s.Close()
+
+	authorizer, err := newV1Authorizer(s.URL, testcerts.ClientCert, testcerts.ClientKey, testcerts.CACert, 0)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	kubeConfig, closeFn := setupWithAuthorizer(t, 10, 10, authorizer)
+	defer closeFn()
+
+	loopbackClient := clientset.NewForConfigOrDie(kubeConfig)
+	noxu1Client := getClientFor(kubeConfig, "noxu1")
+	noxu2Client := getClientFor(kubeConfig, "noxu2")
+
+	queueLength := 50
+	concurrencyShares := 100
+
+	priorityLevelNoxu1, _, err := createPriorityLevelAndBindingFlowSchemaForUser(
+		loopbackClient, "noxu1", concurrencyShares, queueLength)
+	if err != nil {
+		t.Error(err)
+	}
+	priorityLevelNoxu2, _, err := createPriorityLevelAndBindingFlowSchemaForUser(
+		loopbackClient, "noxu2", concurrencyShares, queueLength)
+	if err != nil {
+		t.Error(err)
+	}
+	availableSeats, err := getAvailableSeatsOfPriorityLevel(loopbackClient)
+	if err != nil {
+		t.Error(err)
+	}
+
+	t.Logf("noxu1 priority level concurrency limit: %v", availableSeats[priorityLevelNoxu1.Name])
+	t.Logf("noxu2 priority level concurrency limit: %v", availableSeats[priorityLevelNoxu2.Name])
+	if (availableSeats[priorityLevelNoxu1.Name] <= 4) || (availableSeats[priorityLevelNoxu2.Name] <= 4) {
+		t.Errorf("The number of available seats for test client priority levels are too small: (%v, %v). Expecting a number > 4",
+			availableSeats[priorityLevelNoxu1.Name], availableSeats[priorityLevelNoxu2.Name])
+	}
+
+	stopCh := make(chan struct{})
+	wg := sync.WaitGroup{}
+
+	// "elephant"
+	noxu1NumGoroutines := 5 + queueLength
+	var noxu1ClientRequestLatencySum float64
+	var noxu1ClientRequestLatencySumSq float64
+	var noxu1ClientRequestLatencyCount int32
+	var noxu1Mutex sync.Mutex
+	streamRequestsWithIndex(noxu1NumGoroutines, func(idx int) {
```

Comment:
just curious, what is the `idx` param used for?

Reply:
This is a remnant of a previous version. I will remove `streamReuqestsWithIndex` and use `streamRequests` (defined in concurrency_test.go instead.

Reply:
This is resolved in the new commit